---
title: "RNA-seq differential exon usage"
author: "Michael Hunt"
date: "29 April 2016"
output: html_document
---

[Book page: RNA-seq differential exon usage](http://genomicsclass.github.io/book/pages/rnaseq_exon_usage.html)

## Exon usage assessment

We will look for differential exon usage in the same experimental data as shown in the video, but we will use a different subset of the genes (to speed up the time required for analysis on student's laptops).

Build a DEXSeq dataset object:

```{r}
library("pasilla")
inDir = system.file("extdata", package="pasilla", mustWork=TRUE)        
countFiles = list.files(inDir, pattern="fb.txt$", full.names=TRUE)    
flattenedFile = list.files(inDir, pattern="gff$", full.names=TRUE) 


sampleTable = data.frame(row.names = c( "treated1", "treated2", "treated3","untreated1", "untreated2", "untreated3", "untreated4" ), condition = c("knockdown", "knockdown", "knockdown", "control", "control", "control", "control" ))                 

library("DEXSeq")  
dxd = DEXSeqDataSetFromHTSeq(countFiles, sampleData=sampleTable, 
  design= ~ sample + exon + condition:exon, flattenedfile=flattenedFile )
```

Now we will subset to 1000 genes on chr2L which do not have low counts (just for demontration, typically you would just run the analysis on the whole dataset):

```{r}
rmean = rowMeans(counts(dxd))
```

We use rowRanges to pull out chr2L:

```{r}
dxd2L = dxd[seqnames(rowRanges(dxd)) == "chr2L" & rmean > 10,]
```

Now subset to first 1000:

```{r}
dxd2L = dxd2L[1:1000,]  
```

### Exon usage Q1

What is the gene name of the gene with the exon with the smallest adjusted p-value for differntial exon usage? (The gene name is the part FBgn..., before the ':' and the exon number, E001. So stop before the ':')


```{r}
dxd2L <- DESeq(dxd2L)
res <- results(dxd2L)
head(res)
```


