---
title: "Confounding in Genomics"
author: "Michael Hunt"
date: "April 4, 2016"
output: html_document
---

```{r}
library(Biobase)
library(GSE5859)
data(GSE5859)
```

Note that this is the original dataset from which we selected the subset used in `GSE5859Subset`.  You can obtain it from the genomicsclass GitHub repository.

We can extract the gene expression data and sample information table using the Bio conductor functions `exprs` and `pData` like this:

```{r}
geneExpression = exprs(e)
sampleInfo = pData(e)
```

### Confounding in Genomics Exercises #1  

Familiarize yourself with the sampleInfo table. Note that some samples were processed at different times. This is an extraneous variable and should not affect the values in geneExpression. However, as we have seen in previous analyses it does appear to have an effect so we will explore this here.

You can extract the year from each date like this:

```{r}
year = format(sampleInfo$date,"%y")
```
Note there are

```{r}
length( unique(year) )
```
unique years for which we have data.

For how many of these years do we have more than one ethnicity represented?

```{r}
tab=table(year,sampleInfo$ethnicity)
print(tab)
x=rowSums(tab!=0)
sum(x>=2)
```

### Confounding in Genomics Exercises #2

Repeat the above exercise but now instead of year consider the month as well. Specifically, instead of the year variable defined above use:

```{r}
month.year = format(sampleInfo$date,"%m%y")
```
For what __proportion__ of these `month.year` values do we have more than one ethnicity represented?
t
```{r}
mytab=table(month.year,sampleInfo$ethnicity)
print(mytab)
myx=rowSums(mytab!=0)
mean(myx>=2)
```
Note that this implies that month.year and ethnicity are almost completely confounded. This means that it is hard to separate effects due to date from effects due to our outcome of interest.

### Confounding in Genomics Exercises #3  

Perform a t-test (use rowttests) comparing CEU samples processed in 2002 to those processed in 2003. Then use the `qvalue` package to obtain q-values for each gene.
qwr
How many genes have q-values < 0.05?

```{r}
library(qvalue)
library(genefilter)
year = factor( format(sampleInfo$date,"%y") )
index = which(year%in% c("02","03") & sampleInfo$ethnicity=="CEU")
year = droplevels(year[index])
pval = rowttests(geneExpression[ ,index], year)$p.value
qval = qvalue(pval)
sum(qval$qvalue < 0.05)
```
 
What is the estimate of pi0 provided by qvalue:

```{r}
qval$pi0
```
Note that the estimated percentage of genes that are differentially expressed is above 50%. This is one way to show the magnitude of the effect processing date has on the measurements.

### Confounding in Genomics Exercises #3    

Now perform a t-test (use rowttests) comparing CEU samples processed in 2003 to CEU samples processed in 2004. Then use the `qvalue` package to obtain q-values for each gene.

How many genes have q-values < 0.05?

```{r}
library(qvalue)
library(genefilter)
year = factor( format(sampleInfo$date,"%y") )
index = which(year%in% c("03","04") & sampleInfo$ethnicity=="CEU")
year = droplevels(year[index])
pval = rowttests(geneExpression[ ,index], year)$p.value
qval = qvalue(pval)
sum(qval$qvalue < 0.05)
```
Here we confirm the processing date has an effect on our measurements.

### Confounding in Genomics Exercises #5

Now we are going to compare ethnicities as was done in the original publication in which these data were first presented. Use the rowttests function to compare the ASN population to the CEU population. Once again, use the qvalue function to obtain q-values.

How many genes have q-values < 0.05?

```{r}
library(qvalue)
library(genefilter)
ethnicity = factor( sampleInfo$ethnicity)
index = which(ethnicity%in% c("CEU","ASN"))
ethnicity = droplevels(ethnicity[index])
pval = rowttests(geneExpression[ ,index], ethnicity)$p.value
qval = qvalue(pval)
sum(qval$qvalue < 0.05)
```

### Confounding in Genomics Exercises #6

Note that over 80% of genes are called differentially expressed between ethnic groups. However, due to the confounding with processing date, we need to confirm these differences are actually due to ethnicity. This will not be easy due to the almost perfect confounding. However, above we noted that two groups were represented in 2005. Just like we stratified by majors to remove the "major effect" in our admissions example, here we can stratify by year and perform a t-test comparing ASN and CEU, but only for samples processed in 2005.

How many genes have q-values < 0.05?

```{r}
library(qvalue)
library(genefilter)
ethnicity = factor( sampleInfo$ethnicity)
year = factor( format(sampleInfo$date,"%y") )
index = which(ethnicity%in% c("CEU","ASN") & year=="05")
ethnicity = droplevels(ethnicity[index])
pval = rowttests(geneExpression[ ,index], ethnicity)$p.value
qval = qvalue(pval)
sum(qval$qvalue < 0.05)
```
Note the dramatic drop in the number of genes with q-value < 0.05 when we fix the year. However, the sample size is much smaller in this latest analysis which means we have less power:

```{r}
table(sampleInfo$ethnicity[index])
```

### Confounding in Genomics Exercises #7

To provide a more balanced comparison we repeat the analysis but now taking 3 random CEU samples from 2002. Repeat the analysis above but comparing the ASN from 2005 to three random CEU samples from 2002. Set the seed at 3, `set.seed(3)`

How many genes have q-values < 0.05?

```{r}
library(qvalue)
library(genefilter)
ethnicity = factor( sampleInfo$ethnicity)
year = factor( format(sampleInfo$date,"%y") )
indexASN = which(ethnicity%in% c("ASN") & year=="05")
indexCEU = which(ethnicity%in% c("CEU") & year=="02")
set.seed(3)
indexCEU<-sample(indexCEU,3)
index<-c(indexASN,indexCEU)
ethnicity = droplevels(ethnicity[index])
pval = rowttests(geneExpression[ ,index], ethnicity)$p.value
qval = qvalue(pval)
sum(qval$qvalue < 0.05)
```