---
title: "Data Analysis and Integration: 1. CpGIslandShores"
author: "Michael Hunt"
date: "May 1, 2016"
output: html_document
---

Book chapter:[Inference for DNA methylation data](http://genomicsclass.github.io/book/pages/inference_for_DNAmeth.html)

Path name to folder where we have stored the TCGA data that we downloaded from [here](https://github.com/genomicsclass/tcgaMethylationSubset)

```{r}
#mypath="C:/temp/RSpace/PH525rawdata/PH525tcgaMethylationSubset-master"
mypath="H:/RSpace/PH525rawdata/PH525tcgaMethylationSubset-master"
list.files(mypath)
```

load these Bioconductor packages:

```{r}
library(minfi)
library(IlluminaHumanMethylation450kmanifest)
library(IlluminaHumanMethylation450kanno.ilmn12.hg19)
```

### CpGIslandShores Assessment Q1

Read in the sample annotation table:

```{r}
#path="/YourPath"
targets=read.delim(file.path (mypath,"targets.txt"),as.is=TRUE)
```

How many samples are represented in this table?

```{r}
nrow(targets)
```

### CpGIslandShores Assessment Q2

How many samples are from normal colon samples?

```{r}
nrow(targets[targets$Status=="normal" & targets$Tissue=="colon",])
# or
sum(targets$Tissue=="colon" & targets$Status=="normal")
##or look here
table(targets$Tissue,targets$Status)
```

### CpGIslandShores Assessment Q3

For the next question we will read in breast and colon normal samples:

```{r}
index = which( targets$Status=="normal" & targets$Tissue%in%c("colon","breast") )
targets = targets[index,]
```

Now we are ready to read in the data (this will take about 2 minutes):

```{r}
library(minfi)
dat = read.450k.exp(base=mypath,targets = targets, verbose=TRUE)
```

dat includes the raw data. To convert this into an object that includes methylation values, as well as the location of CpGs, we do the following (we show you the class of dat as we transform it):

```{r}
class(dat)
## preprocess the data
dat = preprocessIllumina(dat)
class(dat)
## assign locations to each CpG
dat = mapToGenome(dat)
class(dat)
## precompute methylation values from U and M values
dat = ratioConvert(dat,type="Illumina")
class(dat)
```

Before we start we can create some quality assessment plots. First look at the distribution of each sample:

```{r}
library(rafalib)
mypar(1,1)
##extract methylation values
y = getBeta(dat)
shist(y)
```

Note that the distributions seem similar. Nothing stands out.

We also create an MDS plot to search for outlier samples. The first PC splits the data by tissue as expected and no sample stands out as an outlier.

```{r}
mds = cmdscale( dist(t(y)))
tissue = as.factor(pData(dat)$Tissue)
plot(mds,col=tissue)
```

Now we are ready to use statistical inference to find differentially methylated regions. Let's start by using the limma package to perform a site-by-site analysis.

```{r}
library(limma)
##create design matrix
tissue = as.factor(pData(dat)$Tissue)
X = model.matrix(~tissue)
##extract methylation values
y = getBeta(dat)
## obtain effect sizes and pvals with limma
fit = lmFit(y,X)
```

Find the CpG with the largest effect size when comparing the two tissues. What chromosome is it on?

```{r}
maxes<-which.max(fit$coefficients[,2])
granges(dat)[maxes]

# model answer

index = which.max(abs( fit$coef[,2]))
seqnames(dat)[index]
start(dat)[index]
```

### CpGIslandShores Assessment Q4

Now we will use the qvalue function to determine the q-value for the CpG found in the previous question.

```{r}
library(qvalue)
##create design matrix
tissue = as.factor(pData(dat)$Tissue)
X = model.matrix(~tissue)
##extract methylation values
y = getBeta(dat)
## obtain effect sizes and pvals with limma
fit = lmFit(y,X)
eb = ebayes(fit)
## obtain q-values
qvals = qvalue(eb$p.value[,2])$qvalue
```

What is the q-value for this CpG?

```{r}
##When reporting such small p-values there is no need to show exactly how small it is
## so we can say < 10^-6
index = which.max(abs( fit$coef[,2]))
qvals[index]
```

### CpGIslandShores Assessment Q5

For this problem, we will use the location of the CpG discussed in the previous two questions. 

Find all the CpGs within 5000 basepairs of the location of this CpG. 

Create a plot showing the methylation values for all samples for these CpGs. Use color to distinguish breast from colon.


Based on these plots, which best describes this region?

 There are differences between colon and breast for all CpGs.  
 No CpGs appear to be different here.  
 A region of about 1000 base pairs appears to be different.  --  correct
 All the CpGs are unmethylated
 
```{r}
# model answer

library(rafalib)
mypar(3,1)
index = which.max(abs( fit$coef[,2]))
gr=granges(dat)[index]+5000
index=which(granges(dat)%over%gr)
pos= start(dat)[index]

matplot(pos,y[index,],ylab="Methylation",col=as.numeric(tissue))

plot(pos, fit$coef[index,2],ylab="Effect Size")

plot(pos, -log10(qvals[index]) ,ylab="-log10 q-value")

```
 
### CpGIslandShores Assessment Q6

Repeat the above exercise, but now make the same plots for the top 10 CpGs ranked by absolute value of effect size. You can get the order like this:

```{r}
library(rafalib)
mypar(3,1)
o = order(abs(fit$coef[,2]), decreasing = TRUE)[1:10]
for(i in o){
  index = i
  gr=granges(dat)[index]+5000
  index=which(granges(dat)%over%gr)
  pos= start(dat)[index]

  matplot(pos,y[index,,drop=FALSE],ylab="Methylation",col=as.numeric(tissue))

  plot(pos, fit$coef[index,2],ylab="Effect Size")

  plot(pos, -log10(qvals[index]) ,ylab="-log10 q-value")
}
```

How would you describe these plots?

 They show only the one CpG in the top 10 as differentially methylated.  
 For most plots we see groups of CpGs that are differentially methylated.  -- correct
 We can hardly see differences between the two tissues.  
 The between tissue variability is too large.
 
### CpGIslandShores Assessment Q7

Now we are going to explicitly search for regions using the bumphunter function. We will use permutation to assess statistical significance. Because the function is slow, we will restrict our analysis to chromosome 15.

```{r}
index= which(seqnames(dat)=="chr15")
dat2 = dat[index,]
```

If your computer has more than one core, you can use parallel computing to speed up the procedure.

```{r}
library(doParallel)
ncores = detectCores()
registerDoParallel(cores = ncores)
```

We can now run the bumphunter function to find differentially methylated regions (DMR). For this assessment question we will use 100 permutations, although we recommend more in practice. Here we will use a cutoff of 0.1. The permutations are random so make sure you set seed to 1 to obtain exact results in the assessment question.

```{r}
##create design matrix
tissue = as.factor(pData(dat)$Tissue)
X = model.matrix(~tissue)
##extract methylation values
set.seed(1)
res = bumphunter(dat2,X,cutoff=0.1,B=100)
head(res$tab)
```

According to these results, how many regions achieve an FWER lower than 0.05?

```{r}
sum(res$table$fwer < 0.05)
```

### CpGIslandShores Assessment Q8

Previously we performed a CpG by CpG analysis and obtained qvalues. Create an index for the CpGs that achieve qvalues smaller than 0.05 and a large effect size larger than 0.5 (in absolute value):

```{r}
##fit and qvals were defined in a previous answer
index = which(qvals < 0.05 & abs(fit$coef[,2]) > 0.5 & seqnames(dat)=="chr15")
```

Now create a table of the DMRs returned by bumphunter that had 3 or more probes and convert the table into GRanges:

```{r}
tab = res$tab[ res$tab$L >= 3,]
tab = makeGRangesFromDataFrame(tab)
```

What proportion of the CpGs indexed by index are inside regions found in tab (hint use the findOverlaps function in the GenomicRanges package)?

```{r}
length(findOverlaps(dat[index,],tab))/length(index)
#or (model answer:
mean(granges(dat[index,])%over%tab)
```

### CpGIslandShores Assessment Q9


Now download the table of CGI using AnnotationHub.

```{r}
library(AnnotationHub)
cgi = AnnotationHub()[["AH5086"]]
```

Now we create a GRanges object from the list of DMRs we computed in the previous questions:

```{r}
tab = res$tab[res$tab$fwer <= 0.05,]
tab = makeGRangesFromDataFrame(tab)
```

What proportion of the regions represented in tab do not overlap islands, but overall CpG islands shores (within 2000 basepairs) ? Hint: use the distanceToNearest

```{r}
map = distanceToNearest(tab,cgi)
d = mcols(map)$distance
mean(d>0 & d<=2000)
```

### CpGIslandShores Assessment Q10

Here we are going to study the relationship between gene expression and DNA methylation by integrating gene expression and DNA methylation high throughput data. We will compare colon and lung samples.

```{r}
path=mypath
targets=read.delim(file.path (path,"targets.txt"),as.is=TRUE)
index = which( targets$Status=="normal" & targets$Tissue%in%c("colon","lung") )
targets = targets[index,]
```

Now we are ready to read in the data (this will take about 2 minutes):

```{r}
library(minfi)
dat = read.450k.exp(base=path,targets = targets, verbose=TRUE)
```

We convert this data as previously done.

```{r}
## preprocess the data
dat = preprocessIllumina(dat)
dat = mapToGenome(dat)
dat = ratioConvert(dat,type="Illumina")
```

Now run the bumphunter function with cutoff=0.25 and default parameters.

```{r}
library(bumphunter)
tissue=pData(dat)$Tissue
X = model.matrix(~tissue)
res = bumphunter(dat,X,cutoff=0.25)
nrow(res$tab)
```
  
What proportion of regions are just one CpG?

```{r}
tab<-makeGRangesFromDataFrame(res$tab,keep.extra.columns=TRUE)
hits<-subjectHits(findOverlaps(dat,tab))
sum(table(hits)==1)/nrow(res$tab)

#or (model answer)
mean(res$tab$L==1)
```

so the L column of the table is the number of CpGs in the region.

```{r}
hist(res$tab$L)
```


### CpGIslandShores Assessment Q11

Now we are going to match these regions to genes. First load the expression related data. Note that this will load an object called tcgaLungColonExpLM which is the result of a differential expression analysis using limma on TCGA rawdata:

```{r}
path=mypath
load(file.path(path,"tcgaLungColonExpLM.rda"))
class(tcgaLungColonExpLM)
```

Note that we saved the annotation of the gene expression array in this object:

```{r}
print( annotation )
```
  
We can obtain q-values using the qvalue package.

```{r}
library(limma)
library(qvalue)
eb=ebayes(tcgaLungColonExpLM)
qvals=qvalue(eb$p.value[,2])$qvalue
```

Now we have to obtain locations for these genes. We will use one of the techniques we learned in course 4. Note: you will need to install the hgu133plus2.db and Homo.sapiens packages from Bioconductor if you have not already.

```{r}
library(hgu133plus2.db)
map=select(hgu133plus2.db,keys=rownames(tcgaLungColonExpLM$coef),columns=c("ENTREZID"),keytype="PROBEID")

library(Homo.sapiens)
Genes=genes(Homo.sapiens)
Genes=resize(Genes,1) ## we want the tss

index1=match(as.character(mcols(Genes)$GENEID),map$ENTREZID)
index2 = match(map$PROBEID[index1],rownames(tcgaLungColonExpLM$coef))
M = tcgaLungColonExpLM$coef[index2,2]
```

M is now in the same order as Genes. We can now find the closest gene to each DMR.

```{r}
tab=makeGRangesFromDataFrame(res$tab,keep.extra.columns = TRUE)
map2=distanceToNearest(tab,Genes)
```

Now we can make plots comparing the methylation differences to the gene expression differences. We consider DMRs of different size

```{r}
index1=subjectHits(map2)
dist = mcols(map2)$dist

library(rafalib)
mypar(2,2)
for(i in c(0,1,2,3)){
  keep = dist< 10000 & tab$L>i
  plot(tab$value[keep],M[index1][keep],main=paste("cor=",signif(cor(tab$value[keep],M[index1][keep],use="complete"),2)))
}
```

What does this figure show?

 There is a positive correlation between gene expression and DNA methylation  
 There is a negative correlation between gene expression and DNA methylation and it is stronger for larger DMRs  -- correct
 There is a netgative correlation between gene expression and DNA methylation and no relation ship to size of DMR  
 There is a no correlation between gene expression and DNA methylation
 
The L column in tab gives the size of the DMR. tab$value gives the methylation value, while M gives the gene expression value. The plots, for L=1,2,3,4 are of gene expression value against methylayion value.