---
title: "Data Analysis and Integration: 2. Cell Composition"
author: "Michael Hunt"
date: "May 4, 2016"
output: html_document
---

[Book chapter: Multidimensional scaling](http://genomicsclass.github.io/book/pages/mds.html) from 4x is helpful here.

### Cell Composition Assessment Q1


To examine the importance of accounting for cellular composition in DNA methylation analysis, we are going to download a GEO dataset used in an analysis of whole blood data.

The minfi package has a function to read data directly from GEO. Run the following commands. Note that this command downloads 64.7 MB of data and can take several minutes, depending on your download speed.

```{r}
library(minfi)
grset=getGenomicRatioSetFromGEO("GSE32148")
```

```{r}
path="C:/Users/Mike/RSpace/PH525rawdata"
#path="H:/RSpace/PH525rawdata/"
load(paste(path,"/grset.rda",sep=""))
```

This creates an object of class:

```{r}
class(grset)
```

which is quite convenient. Use the function pData to examine the sample information table and determine which column includes the age of the individual.

What is the name of this column?

```{r}
names(pData(grset))
pData(grset)[1,]
## note age is here
pData(grset)[1,"characteristics_ch1.1"]
```

### Cell Composition Assessment Q2

With a bit of wrangling we can extract the age as a number of the individuals sampled here:

```{r}
age=pData(grset)$characteristics_ch1.1
age=as.character(age)
age[grep("N/A",age)] = NA
age=as.numeric(gsub("age \\(y\\):\ ","",age))
```

This experiment was performed to find DMRs between

individuals with Crohn's disease and controls. We can

extract this information like this:

```{r}
group = rep("normal",nrow(pData(grset)))
group[grepl("ulcerative",pData(grset)[,1])]="ulcerative"
group[grepl("Crohn",pData(grset)[,1])]="crohn"
group = factor(group,levels=c("normal","ulcerative","crohn"))
```

We want to create some exploratory plots based on distance between sample. Before doing this, we will remove CpGs with NA calls as well as the sex chromosomes:

```{r}
keep = which(rowSums(is.na(getBeta(grset)))==0 & 
               !seqnames(grset)%in%c("chrX","chrY"))
##create a new object 
grset2=grset[keep,]
```

Use a multi-dimensional scaling (MDS) plot on the methylation measurements in grset2. Use color to denote individuals that are older than 40 and different symbols to denote the different conditions: normal, ulcerative, crohn.

```{r}
y = getBeta(grset2)
mds = cmdscale( dist( t(y)))
mypar(1,1)
col=ifelse(age>40,1,2)
plot(mds,pch=as.numeric(group),col=col)
legend("bottomleft",levels(group),pch=seq_along(group))
```

How would you describe the MDS-plot?

 Samples cluster by condition, normals to the left and others to the right. 
 There is no obvious clustering.  
 The individuals that are older than 40 form a cluster, with the rest showing larger variability.  
 The samples cluster perfectly by age. 