---
title: "RNA-seq alignment"
author: "Michael Hunt"
date: "April 26, 2016"
output: html_document
---

## FASTQ

Here is the beginning of the file SRR1039508_1.fastq, which contains the first read of the pair for the experiment:

@SRR1039508.1 HWI-ST177:290:C0TECACXX:1:1101:1225:2130/1
CATTGCTGATACCAANNNNNNNNGCATTCCTCAAGGTCTTCCTCCTTCCCTTACGGAATTACA
+
HJJJJJJJJJJJJJJ########00?GHIJJJJJJJIJJJJJJJJJJJJJJJJJHHHFFFFFD
@SRR1039508.2 HWI-ST177:290:C0TECACXX:1:1101:1311:2131/1
CCCTGGACTGCTTCTTGAAAAGTGCCATCCAAACTCTATCTTTGGGGAGAGTATGATAGAGAT
+
HJJJJJJJJJJJJJJJJJIIJIGHIJJJJJJJJJJJJJJJJJJJJJJGHHIDHIJJHHHHHHF
@SRR1039508.3 HWI-ST177:290:C0TECACXX:1:1101:1414:2163/1
TCGATCCATCGATTGGAAGGCACTGATCTGGACTGTCAGGTTGGTGGTCTTATTTGCAAGTCC
+
HJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJGJJIGHIJJBGIJCGIAHIJHHHHHHHFFFFF

Here we see three reads, with names, sequence, +, then quality score.

We will try to find the position of the third read using a simple BLAST search to the human genome (build hg19 / GRCh37).

On the [BLAST search](http://blast.ncbi.nlm.nih.gov/Blast.cgi?PAGE_TYPE=BlastSearch&BLAST_SPEC=OGP__9606__9558&LINK_LOC=blasthome) page, under Choose Search Set > Database > choose Genome (GRCh37.p13 reference assembly...)

### FASTQ Q1

The third read aligns uniquely to chromosome 16 of the human genome, falling in the second exon of the gene DHX38, which is on the plus strand (transcribed from left to right). What is the genomic position to which the first basepair of the third read aligns? (Use human genome build: hg19 / GRCh37)

Answer:  
Under "Query Sequence" enter

>SRR1039508.3 HWI-ST177:290:C0TECACXX:1:1101:1414:2163/1
TCGATCCATCGATTGGAAGGCACTGATCTGGACTGTCAGGTTGGTGGTCTTATTTGCAAGTCC

Get: 72130081

Searching with BLAST we get a unique hit, on the plus strand, to chromosome 16:

Range 1: 72130081 to 72130143

We could also use Bioconductor:

```{r,message=FALSE}
# note the difference to the next chunk - here we are reading a plus strand.
library(BSgenome.Hsapiens.UCSC.hg19)
# matchPattern is from library Biostrings
matchPattern("TCGATCCATCGATTGGAAGGCACTGATCTGGACTGTCAGGTTGGTGGTCTTATTTGCAAGTCC", Hsapiens$chr16)
```

### FASTQ Q2

The first three reads in the second file are as follows:

@SRR1039508.1 HWI-ST177:290:C0TECACXX:1:1101:1225:2130/2
CAGATGAGGCGTGTTGGCCAGAGAGCCATTGTCAACAGCAGAGATGNNNNNNNNNNNNAATCC
+
HJJJJJJJJJJHIIIJJJJJJJJJJJJJJJJJJJJJJJHIJIJHII#################
@SRR1039508.2 HWI-ST177:290:C0TECACXX:1:1101:1311:2131/2
TACTCCGGAGAACAGATGGGATTCCCTAGGAGACCCTTGAGGGAAAAGGGAGCCCCAATCTCT
+
FJJJJJJJFHEHJJJHIIJJGGIIJJGIIJGJHJJJJJHGIJJIGIHHHHFFFDDDDDDDDDE
@SRR1039508.3 HWI-ST177:290:C0TECACXX:1:1101:1414:2163/2
TCGCTCTCTCCGTTTCAGGGAAGCCAGCAAGTCCAGTCCGAGTAATGAAGGGCGGGGAGCAGG
+
HJJJJJJJJJJJJJJJJJJJIJJJJJJJJJJIJJJJJJFHIJFHJJJJJIHHHFDDDDDDDDD

Align the third read above to the human genome, as we did previously.

The third read in the second file also aligns uniquely to chromosome 16 of the human genome, aligning to the _minus_ strand, also in the second exon of the gene DHX38. What is the genomic position to which the first basepair of the third read in the second file aligns?

Answer:  72130242  
Searching with BLAST we get a unique hit, on the minus strand, to chromosome 16:

Range 1: 72130180 to 72130242

So the first basepair of the read is the last basepair in the alignment: 72130242

We could also use Bioconductor:

```{r,message=FALSE}
# note the difference to the previous chunk - here we are reading a minus strand.
library(BSgenome.Hsapiens.UCSC.hg19)
matchPattern(reverseComplement(DNAString("TCGCTCTCTCCGTTTCAGGGAAGCCAGCAAGTCCAGTCCGAGTAATGAAGGGCGGGGAGCAGG")), Hsapiens$chr16)
```

Above we saw the two reads of a paired-end fragment, both aligning within an exon of a human gene.

For a strand-specific RNA-sequencing protocol, we could have seen this kind of alignment to a + strand gene:

[1st read + strand] ... [2nd read - strand]

...or we could have seen the 1st read aligning on the right side:

[2nd read + strand] ... [1st read - strand]

A strand-specific protocol means that we only observe fragments with the same strand as the gene.

However, many experiments, including the one we are examining are not strand-specific. This means that, for a plus strand gene, we will observe, in addition to the above two kinds of paired-end fragments, two more:

[1st read - strand] ... [2nd read + strand]

[2nd read - strand] ... [1st read + strand]

## FASTQC

Unzip and open the FASTQC report for the second file. The following questions will ask about the information in this report. Remember you can check the FASTQC help files for more information (in particular the Analysis Modules explanations).

### FASTQC Q1

Which basepair has the lowest 10% quality score?

9. In the second module, 'Per base sequence quality', the whiskers on the boxplots indicate the 10% and 90% of the quality scores at each position.


### FASTQC Q2

What base is most common in the first position of the read?

T. The red line is highest at position 1 in the fourth module, "Per base sequence content".

### FASTQC Q3

The sequence GGAAA spikes at position 10, what is the observed / expected value at this position for GGAAA?

5.1476827. The last module, "K-mer content", has a table with the observed / expected for the position with the max.

## Genome alignment 

The header of the Aligned.out.sam file is pasted below:

```
@PG     ID:STAR PN:STAR VN:STAR_2.3.1z_r395     CL:STAR   --runThreadN 12   --genomeDir GRCh38.79.chrom1   --readFilesIn fastq/SRR1039508_1.fastq   fastq/SRR1039508_2.fastq    cl:STAR --runThreadN 12 --genomeDir GRCh38.79.chrom1 --readFilesIn fastq/SRR1039508_1.fastq fastq/SRR1039508_2.fastq
@SQ     SN:1    LN:248956422
SRR1039508.7    163     1       96446974        255     63M     =       96446997        86      CGTAGATTCGGGCAAGTCCACCACTACTGGCCATCTGATCTATAAATGCGGTGGCATCGACAA HJHIJJJJJJJJJJJJJJJIJJJJJJJJIJJJJJJJJJIIJJJJJJJJJJJEHHFFFFDEDDD NH:i:1  HI:i:1  AS:i:110        nM:i:7
SRR1039508.7    83      1       96446997        255     63M     =       96446974        -86     CTACTGGCCATCTGATCTATAAATGCGGTGGCATCGACAAAAGAACCATTGAAAAATTTGAGA FHHHHHJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJIJJJJJJJJJJJJJJJJJH NH:i:1  HI:i:1  AS:i:110        nM:i:7
SRR1039508.16   163     1       243426450       255     63M     =       243426495       108     AAGAAAAAAGGTATACATATGATAAATTGGGAAAGTTACAGAGAAGAAATGAAGAATTGGAGG HJJJJJJJJJJAGIJJJIJJJJJJJJJJIJJIIJJHIJFHJIJJIJIJJJJJJJJIJJHHHHG NH:i:1  HI:i:1  AS:i:124        nM:i:0
SRR1039508.16   83      1       243426495       255     63M     =       243426450       -108    GAAATGAAGAATTGGAGGAACAGTGTGTCCAGCATGGGAGAGTACATGAGACGATGAAGCAAA IJJIGJJJJJIIIJJJJJIIJJJJJJJIJJJJJIGJJIJJJJJIJJJJIJJJJJJJJJJJJJH NH:i:1  HI:i:1  AS:i:124        nM:i:0
```

The [SAM](https://samtools.github.io/hts-specs/SAMv1.pdf) format and its compressed version, BAM, contains information about the alignment of sequences to a reference genome.

Section 1.3 describes the header of the SAM file.


### Genome align Q1

What is the length of reference sequence 1 according to our SAM file?
From checking the table in section 1.3 of the SAM format specification, we know the @SQ gives the reference sequence 1 length with LN:

@SQ SN:1 LN:248956422

### Genome align Q2

In section 1.4 on the alignments, the different fields are described. The CIGAR string tells us how the read aligned to the genome, using numbers and letters, M for match, I for insertion, D for deletion, N for skipped bases on the reference genome.

What is the CIGAR string for all of the two pairs of reads we are shown above? (The answer includes numbers and a letter)

63M.The table in section 1.4 shows that the CIGAR string is in the 6th column. All reads have 63M here.

### Genome align Q3

In section 1.4, we see that the second column is a bitwise FLAG, an integer which encodes possible YES/NO answers to the choices below.

The bitwise FLAG is most easily decoded at this link, [Explain SAM Flags](http://broadinstitute.github.io/picard/explain-flags.html), for the PICARD software.

Type the bitwise FLAG of the top read above into the SAM flag explainer. Which choices are true about this read?

 read paired  -- correct  
 read mapped in proper pair -- correct   
 read unmapped  
 mate unmapped  
 read reverse strand  
 mate reverse strand  -- correct  
 first in pair  
 second in pair  -- correct  
 not primary alignment  
 read fails platform/vendor quality checks  
 read is PCR or optical duplicate  
 supplementary alignment
 
 Typing '163' into the SAM flag explainer gives us the right choices above.
 
 Read 7 in the first FASTQ file was:

```
@SRR1039508.7 HWI-ST177:290:C0TECACXX:1:1101:1952:2109/1
TCTCAAATTTTTCAATGGTTCTTTTGTCGATGCCACCGCATTTATAGATCAGATGGCCAGTAG
+
HJJJJJJJJJJJJJJJJJIJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJHHHHHF

Read 7 in the second FASTQ file was:

@SRR1039508.7 HWI-ST177:290:C0TECACXX:1:1101:1952:2109/2
CGTAGATTCGGGCAAGTCCACCACTACTGGCCATCTGATCTATAAATGCGGTGGCATCGACAA
+
HJHIJJJJJJJJJJJJJJJIJJJJJJJJIJJJJJJJJJIIJJJJJJJJJJJEHHFFFFDEDDD
```

The '163' flag tells us that the first read in the SAM files corresponds to read 7 from the second FASTQ file. Note that the sequence of the read in the SAM file starts with "CGTAGA...", as does read 7 in the second FASTQ file. It also tells us that the mate is reverse strand. Note that read 7 in the first FASTQ file starts with "TCTCAA...", while the second read in the SAM file above ends with "...TTGAGA".

In the FASTQ file, the sequence of the reads was from opposite strands, while in the SAM file, the sequence for the reads is on the same strand, that is, the strand has been reversed for one of the pairs.

## Integrative Genomics Viewer (IGV)

__NOTE__: in the previous video, we aligned reads to the most current Ensembl reference genome, hg38 / GRCh38. In this video, and in the sections of Week 2, we will examine reads in an already prepared BAM file, though this file was prepared more than a year ago. The reads in the BAM files provided were aligned to an earlier build of the human genome: __hg19 / GRCh37__. This is a good point to remember: that, if you receive an alignment file (BAM file), you should be confident that you know to which genome the reads were aligned. This is critical information for you to visualize and properly analyze the data.

__NOTE__: in the video, alignments may appear earlier while zooming in than on your machine. There is a setting which controls this: View > Preferences > Alignments > __Visibility range threshold__. (Thanks to Fabio Amaral for pointing this out). So don't be surprised if the alignments in the video appear while on your machine it says "Zoom in to view the alignments".

[IGV software homepage: ](https://www.broadinstitute.org/igv/) 

The BAM file used in this video can be downloaded [here](https://courses.edx.org/c4x/HarvardX/PH525.5x/asset/SRR1039508_subset.bam) (less than 1 Mb). Note that this contains reads aligned to hg19.  

The BAM Index file used in this video can be downloaded [here](https://courses.edx.org/c4x/HarvardX/PH525.5x/asset/SRR1039508_subset.bam.bai)
.
In this assessment we will use IGV to visualize the BAM file (reads aligned to hg19) seen in the previous video (which requires its index file as well).

Once you have started IGV, you may have to click "More..." under genomes to get hg19.

Zoom in to gene SRM which is within the range chr1:11,112,000-11,122,000.

### IGV Q1

A giveaway: which strand is this gene: plus or minus?

minus., 
RNA-seq > RNA-seq alignment > IGV assessment
Previous Unit
 other Links for RNA-seq alignment
Note on this subsection 	 video Note
Downloading FASTQ files 	 video First look at a FASTQ file 	 problem FASTQ assessment
FASTQ Q1
Second read file
FASTQ Q2
Strand-specific protocol or not strand-specific 	 video Quality control with FASTQC
Links 	 other FASTQC note
Other processing options 	 problem FASTQC assessment
FASTQC Q1
FASTQC Q2
FASTQC Q3 	 video Note
Genome alignment with STAR I 	 video Genome alignment with STAR II
Downloads 	 problem Genome alignment assessment
SAM format
Genome align Q1
Genome align Q2
SAM FLAGS
Genome align Q3
Checking the FASTQ sequence 	 video Note
Integrative Genomics Viewer (IGV)
Links 	 problem IGV assessment
IGV Q1
IGV Q2
Known SNP?
IGV Q3 	 video Transcriptome alignment with RSEM I
Links 	 video Transcriptome alignment with RSEM II
Downloads 	 problem Transcriptome alignment assessment
Quick data cleaning
Transcript align Q1
Transcript align Q2
Transcript align Q3
Number of isoforms per gene
FPKM and TPM 
Next Unit
 Click to add  Bookmark
In this assessment we will use IGV to visualize the BAM file (reads aligned to hg19) seen in the previous video (which requires its index file as well).

Once you have started IGV, you may have to click "More..." under genomes to get hg19.

Zoom in to gene SRM which is within the range chr1:11,112,000-11,122,000.

### IGV Q1

(1/1 point)
A giveaway: which strand is this gene: plus or minus?

minus - correct

The arrows in the introns of the RefSeq Genes track point to the left.

### IGV Q2

In the first exon (rightmost) there is an alternate allele (C) to the reference (T). What is the genomic location of this SNP? (The number of the basepair only, don't include the chromosome.)

11119899. The SNP is visible as a blue vertical line when zoomed out, upon zooming in we can see the alternate allele, the reference allele and the location from hovering over the position.

Look up the region around the SNP: chr1:11,119,891-11,119,910 in the UCSC Genome Browser (hg19).

You can right click and "Show details" for objects in the UCSC Genome Browser.

This appears to be a known and synonymous variant (meaning it does not change the amino acid sequence in the resulting protein because both codons code for the same amino acid: S (TCA) --> S (TCG).

### IGV Q3

Back in IGV, right click on the BAM file track and choose Squished. We want to count how many reads are in a particular intron. First we will make sure that we see all the possible reads.

Click View > Preferences > Alignments tab. At the top, make sure that "Downsample reads" is NOT checked. Later if you are using IGV for your own work, you may want to check this again, as it speeds up the visualization. But for this question we want to see all the reads.

How many reads are in the first intron (the rightmost intron)? (Just count by eye the number of reads falling entirely in this intron.)

3. While the RNA-seq protocol is highly enriched for spliced mRNA, it is not uncommon to see a few reads in the intron or inter-genic regions. In Week 2, for analyzing expression, and differences in expression across samples, we will not count these reads.

## Transcriptome alignment with RSEM I

RSEM software homepage: http://deweylab.biostat.wisc.edu/rsem/

RSEM paper: http://www.ncbi.nlm.nih.gov/pmc/articles/PMC3163565/

The commands used in this video are provided in a Markdown file [here](https://github.com/genomicsclass/labs/blob/master/rnaseq/trancsript_align_RSEM.md).

In the previous video we calculated expression by probabilistically aligning reads with the RSEM software. And we have provided for you two results files:

genes.results (.5 Mb)

isoforms.results (1 Mb)

We will investigate the values for various columns in these tables. For reference, the documentation for these file formats is here.

First, read these into an R session with the following lines:

```{r}
genes <- read.table("SRR1039508.genes.results", header=TRUE)
isoforms <- read.table("SRR1039508.isoforms.results", header=TRUE)
```

Both 'genes' and 'isoforms' have a column FPKM. 

Using split() and sapply(), confirm that the FPKM column in 'genes' is the sum of the FPKM column in 'isoforms':

```{r}
fpkm.per.gene <- split(isoforms$FPKM, isoforms$gene_id)
head(sapply(fpkm.per.gene, sum))
head(genes$FPKM)
```

Or equivalently using dplyr:

```{r}
# if not yet installed: install.packages("dplyr")
library(dplyr)
isoforms %>% group_by(gene_id) %>% summarize(sum = sum(FPKM))  
```

 

Make a histogram of the FPKM column in 'genes'. Make a histogram after transforming by: log10(x+1).

```{r}
hist(genes$FPKM,breaks=50)
hist(log10(genes$FPKM+1),breaks=50)
```

Loads of genes with FPKM = 0

Make new tables, removing the genes with FPKM 0:

```{r}
genes2 <- genes[genes$FPKM > 0,]
genes2$gene_id <- droplevels(genes2$gene_id)
isoforms2 <- isoforms[isoforms$gene_id %in% genes2$gene_id,]
isoforms2$gene_id <- droplevels(isoforms2$gene_id)
```

We will continue below using these new tables.

We perform a check that the "gene_id" column in genes2 is equal to the levels of the "gene_id" factor in isoforms2:

```{r}
stopifnot(all(genes2$gene_id == levels(isoforms2$gene_id)))
```

If this line runs without error, then we know the check passed.

With 'genes2', make a plot of the effective_length (x-axis) and the expected_count (y-axis), with both axes on the log scale: log="xy".

```{r}
plot(genes2$effective_length,genes2$expected_count,log="xy")
```

### Transcript align Q1

Make a histogram of the FPKM in genes2. Make a histogram after transforming with: log10(x).

What is the median FPKM of the genes with positive FPKM (i.e. the genes2 table)?


```{r}
hist(genes2$FPKM,breaks=50)
hist(log10(genes2$FPKM+1),breaks=50)
summary(genes2$FPKM)
```

### Transcript align Q2

With 'isoforms2', calculate the maximum IsoPct for each gene, 'max.iso'. Make a histogram of 'max.iso'.

For what ratio of genes is the maximum IsoPct higher than 95? (Enter a number between 0 and 1.)

```{r}
# if not yet installed: install.packages("dplyr")
library(dplyr)
res<-isoforms2 %>% group_by(gene_id) %>% summarize(max.iso = max(IsoPct))
hist(res$max.iso)
mean(res$max.iso>95)
```

model answer:
```{r}
# with dplyr:
isoforms2 %>% group_by(gene_id) %>% summarize(max.iso = max(IsoPct)) %>% summarize(answer = mean(max.iso > 95))

# and without: 
max.iso = sapply(split(isoforms2$IsoPct, isoforms2$gene_id), max)

mean(max.iso > 95)
```

### Transcript align Q3

Is there a relationship between total expression and the maximum IsoPct per gene?

Make a plot of 'max.iso' on the x-axis and genes2$FPKM on the y-axis (with log="y").

```{r}
plot(res$max.iso,genes2$FPKM,log="y")
```

Sometimes it is difficult to see relationships when many points stack at one value. Try a boxplot:

```{r}
boxplot(split(log10(genes2$FPKM), cut(max.iso, 5)), xlab="max.iso", ylab="log10 FPKM")
```

Which group has the lowest median FPKM? Genes with 'max.iso' in the range:

 12.5 - 30.1  
 30.1 - 47.6  
 47.6 - 65  
 65 - 82.5  
 82.5 - 100  - correct
 
Calculate the number of isoforms per gene, and plot the maximum IsoPct over the number of isoforms:

```{r}
num.iso <- as.numeric(table(isoforms2$gene_id))
plot(num.iso, max.iso)
```

Make a barplot of the number of genes according to number of isoforms:

```{r}
barplot(table(num.iso))
barplot(table(num.iso[num.iso < 15]))
```

The FPKM (fragments per kilobase of sequence per million mapped reads) is a normalized measure of expression level (having divided out the transcript length and the number of mapped reads).

TPM (transcripts per million), is a linear scaling of the FPKM, such that we would expect a gene with 1 TPM to have one molecule in a population of one million mRNA.

With genes2 plot the TPM over the FPKM.

```{r}
plot(genes2$FPKM,genes2$TPM)
```

Confirm that TPM is equal to: (FPKM / sum(FPKM)) * 1 million

```{r}
plot(genes2$TPM,1e6*genes2$FPKM/sum(genes2$FPKM))
abline(0,1)
```
